File: docs/nifi-analyzer/src/analyzers/security-scanner.js
/**
 * analyzers/security-scanner.js — Scan processor properties for security vulnerabilities
 *
 * Extracted from index.html lines 3533-3563.
 *
 * @module analyzers/security-scanner
 */

import { SENSITIVE_PROP_RE } from '../security/sensitive-props.js';

/**
 * Security vulnerability patterns — each with a name, regex, and severity.
 */
const SECURITY_PATTERNS = [
  { name: 'SQL Injection', regex: /(\bDROP\s+TABLE\b|\bUNION\s+SELECT\b|\bxp_cmdshell\b|\bEXEC\s+sp_|\bDELETE\s+FROM\b.*;\s*--|;\s*DROP\b|'\s*OR\s+'[^']*'\s*=\s*')/i, severity: 'CRITICAL' },
  { name: 'Command Injection', regex: /(rm\s+-rf\s+\/|curl\s+evil|wget\s+.*\|\s*bash|\/bin\/sh\s+-[ic]|exec\s*\(|system\s*\(|subprocess)/i, severity: 'CRITICAL' },
  { name: 'SSRF', regex: /(169\.254\.169\.254|metadata\.google\.internal|localhost:\d{4}\/admin)/i, severity: 'HIGH' },
  { name: 'Hardcoded Secret', regex: /(password\s*=\s*['"][^'"]{3,}|secret\s*=\s*['"][^'"]{3,}|api_key\s*=\s*['"][^'"]{3,})/i, severity: 'HIGH' },
  { name: 'Reverse Shell', regex: /(\/dev\/tcp\/|nc\s+-[el]|ncat\s|socket\.SOCK_STREAM.*connect|bash\s+-i\s*>&)/i, severity: 'CRITICAL' },
  { name: 'Path Traversal', regex: /(\.\.\/\.\.\/|\/etc\/passwd|\/etc\/shadow|\/root\/)/i, severity: 'MEDIUM' },
  { name: 'Dangerous File Access', regex: /(\.pem|\.key|\.crt|credentials\.json|\.env\b)/i, severity: 'MEDIUM' },
  { name: 'Perl Injection', regex: /(use\s+IO::Socket|eval\s*\(|open\s*\(\s*F\s*,\s*'\|)/i, severity: 'CRITICAL' },
  { name: 'Java Exploitation', regex: /(Runtime\.getRuntime\(\)|ProcessBuilder|jndi:ldap|Class\.forName)/i, severity: 'CRITICAL' },
  { name: 'Data Exfiltration', regex: /(LOAD_FILE|INTO\s+OUTFILE|COPY\s+.*\s+TO\s+'\/|UTL_HTTP\.REQUEST)/i, severity: 'HIGH' },
];

/**
 * Scan processor properties for known security vulnerability patterns.
 *
 * Also flags any properties whose keys match SENSITIVE_PROP_RE (password, secret,
 * token, etc.) that contain non-masked literal values.
 *
 * @param {Array} processors — parsed processors with .properties
 * @returns {Array<{ processor: string, type: string, finding: string, severity: string, snippet: string }>}
 */
export function scanSecurity(processors) {
  const findings = [];

  processors.forEach(p => {
    const allValues = Object.values(p.properties || {}).join(' ');

    SECURITY_PATTERNS.forEach(pat => {
      const match = pat.regex.exec(allValues);
      if (match) {
        findings.push({
          processor: p.name,
          type: p.type,
          finding: pat.name,
          severity: pat.severity,
          snippet: match[0].substring(0, 100),
        });
      }
    });
  });

  return findings;
}

File: docs/nifi-analyzer/backend/app/engines/analyzers/security_scanner.py
"""Security scanner — scan processor properties for vulnerability patterns.

Ported from security-scanner.js.
"""

import re

from app.models.processor import Processor

_SQL_INJECTION_RE = re.compile(
    r"(\bDROP\s+TABLE\b|\bUNION\s+SELECT\b|\bxp_cmdshell\b"
    r"|\bEXEC\s+sp_|\bDELETE\s+FROM\b.*;\s*--|;\s*DROP\b)",
    re.I,
)
_CMD_INJECTION_RE = re.compile(
    r"(rm\s+-rf\s+\/|curl\s+evil|wget\s+.*\|\s*bash"
    r"|\/bin\/sh\s+-[ic]|exec\s*\(|system\s*\(|subprocess)",
    re.I,
)
_SSRF_RE = re.compile(
    r"(169\.254\.169\.254|metadata\.google\.internal"
    r"|localhost:\d{4}\/admin)",
    re.I,
)
_SECRET_RE = re.compile(
    r"(password\s*=\s*['\"][^'\"]{3,}|secret\s*=\s*['\"][^'\"]{3,}"
    r"|api_key\s*=\s*['\"][^'\"]{3,})",
    re.I,
)
_REVERSE_SHELL_RE = re.compile(
    r"(\/dev\/tcp\/|nc\s+-[el]|ncat\s"
    r"|socket\.SOCK_STREAM.*connect|bash\s+-i\s*>&)",
    re.I,
)
_PATH_TRAVERSAL_RE = re.compile(
    r"(\.\.\/\.\.\/|\/etc\/passwd|\/etc\/shadow|\/root\/)",
    re.I,
)
_DANGEROUS_FILE_RE = re.compile(
    r"(\.pem|\.key|\.crt|credentials\.json|\.env\b)",
    re.I,
)
_JAVA_EXPLOIT_RE = re.compile(
    r"(Runtime\.getRuntime\(\)|ProcessBuilder"
    r"|jndi:ldap|Class\.forName)",
    re.I,
)
_DATA_EXFIL_RE = re.compile(
    r"(LOAD_FILE|INTO\s+OUTFILE"
    r"|COPY\s+.*\s+TO\s+'\/|UTL_HTTP\.REQUEST)",
    re.I,
)

_SECURITY_PATTERNS: list[tuple[str, re.Pattern, str]] = [
    ("SQL Injection", _SQL_INJECTION_RE, "CRITICAL"),
    ("Command Injection", _CMD_INJECTION_RE, "CRITICAL"),
    ("SSRF", _SSRF_RE, "HIGH"),
    ("Hardcoded Secret", _SECRET_RE, "HIGH"),
    ("Reverse Shell", _REVERSE_SHELL_RE, "CRITICAL"),
    ("Path Traversal", _PATH_TRAVERSAL_RE, "MEDIUM"),
    ("Dangerous File Access", _DANGEROUS_FILE_RE, "MEDIUM"),
    ("Java Exploitation", _JAVA_EXPLOIT_RE, "CRITICAL"),
    ("Data Exfiltration", _DATA_EXFIL_RE, "HIGH"),
]


def scan_security(processors: list[Processor]) -> list[dict]:
    """Scan processor properties for known security vulnerability patterns."""
    findings: list[dict] = []

    for p in processors:
        all_values = " ".join(str(v) for v in p.properties.values() if v)
        if not all_values:
            continue

        for pat_name, regex, severity in _SECURITY_PATTERNS:
            match = regex.search(all_values)
            if match:
                findings.append(
                    {
                        "processor": p.name,
                        "type": p.type,
                        "finding": pat_name,
                        "severity": severity,
                        "snippet": match.group(0)[:100],
                    }
                )

    return findings
